Rewrite `merge.ts` in TypeScript

fnky opened the pull request:
#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [ ] I have included some tests.
- [ ] If it's a new feature, I have included documentation updates.

#### Changes

- Rewrite `merge.ts` in TypeScript.
- Add additional types for hooks for type completion in `known-hook-events.ts`.
- Changes the exported API slightly due to the way they are exported in TypeScript.

fnky commented on source/merge.ts:
```diff
@@ -0,0 +1,93 @@
+import knownHookEvents, {Hooks, HookType, HookEvent} from './known-hook-events';
+
+// TODO: Use the Got options-object types.
+interface Options {
```
The extranous `Options` interface is meant to be merged with the interface that was introduced in https://github.com/sindresorhus/got/pull/708 into a single shared interface.

fnky commented on source/merge.ts:
```diff
@@ -0,0 +1,93 @@
+import {URL} from 'url';
+import is from '@sindresorhus/is';
+// XO doesn't recognize `export type` for some reason.
+// eslint-disable-next-line import/named
```
XO fails at linting for exported `type`s from `known-hook-events.ts`.

fnky replied:
I wasn’t sure we wanted to do this across the project, but it makes it less confusing for future contributors.

fnky commented on source/merge.ts:
```diff
@@ -0,0 +1,93 @@
+	//       populate the `hooks` object in the loop below, TypeScript want us to //       put them into the object upon initialization, because it cannot infer
+	//       that they are going to conform correctly in runtime.
+	// eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
+	const hooks = {} as { [Key in HookEvent]: HookType[] };
```
This one is quirky and probably not correct, but I have tried multiple ways to reflect that we _know_ that it's going to be populated later, in the loop underneath but we don't want to give it the values upfront.

szmarczak replied:
I'm not a TypeScript master either :P

Is there maybe another way? For example the [Hooks interface](https://github.com/sindresorhus/got/pull/714/files#diff-8cbb1059c76170f4b0419c6df7f197baR18):

```ts
const hooks: Hooks = {}
```

szmarczak replied:
ping?

fnky replied:
I have tried that and also with `as Hooks`.

```ts
hooks[hook] = hooks[hook].concat(source.hooks[hook] || []);
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         [ts] Cannot invoke an expression whose type lacks a call signature.
                         Type '{ (...items: ConcatArray<InitHook>[]): InitHook[]; (...items: (InitHook | ConcatArray<InitHook>)[]): InitHook[]; } | { (...items: ConcatArray<BeforeRequestHook>[]): BeforeRequestHook[]; (...items: (BeforeRequestHook | ConcatArray<...>)[]): BeforeRequestHook[]; } | { ...; } | { ...; } | { ...; } | { ...; }'
                         has no compatible call signatures. [2349]
```

Seen as this line is temporary, I don't think its much of an issue just an annoyance.

szmarczak replied:
Oh, thanks for the explaination, it's clear now. I'll try to think of something...

fnky commented on source/merge.ts:
```diff
@@ -0,0 +1,93 @@
+
+type IterateFunction = (options: Options) => void;
+
+type Method = 'GET' | 'PUT' | 'HEAD' | 'DELETE' | 'OPTIONS' | 'TRACE';
```
Maybe these types should be exported from another file to make them available across the codebase as well as the end-user API, if necessary.

fnky commented on source/known-hook-events.ts:
```diff
@@ -1,8 +1,47 @@
+import {RequestOptions} from 'https';
+import {IncomingMessage} from 'http';
+
+// TODO: This may be a bit overkill. It was done for different reasons:
```
I hope you understand the rationale here, even though it does look quite cumbersome.

sindresorhus commented on source/known-hook-events.ts:
```diff
@@ -1,8 +1,47 @@
+	beforeRequest: BeforeRequestHook[];
+	beforeRedirect: BeforeRedirectHook[];
+	beforeRetry: BeforeRetryHook[];
+	afterResponse: AfterResponseHook[];
```
Should the docs be here or on the actual types? For example, line 11.

fnky replied:
This was on oversight on my part. I think I used this to check if you could document each property, and I think you can. I’ll suggest we do it on the properties as well as the types, since both can be consumed by different means.

fnky replied:
Sorry I got it confused with on the type of `Hooks` vs. the properties that takes the list of hooks.

I tried to add the documentation at the type level, but TypeScript/VSCode apparently can't show proper intellisense when you type a function in the array. I think it makes sense to add them to the properties in that case.

fnky replied:
So the consensus is to add the same documentation for both types and the properties in Hooks?

sindresorhus replied:
Yup

fnky commented:
I have pushed a few changes regarding the review.

- Moved the `Options` interface into `utils/types.ts` so it can be reused.
- Added a new `CancelableRequest` type which extends from `PCancelable` with additional properties.
- Fixed some types that were incorrect.
- Added documentation for `Hooks` and its properties.

There are still some things that are needed to make it closer to what is documented:

##### `beforeError` error types

The error passed to the `beforeError` hook can be a number of different types of errors with different properties. I initially tried to make a `AnyError` type that was a union of these different errors, but I'll wait for #712 to be merged so we can discuss any concerns regarding what errors/properties are missing.

##### Extended Response object

There are multiple places where it is documented that you can retrieve for example `.body` from the response. Since `IncomingMessage` doesn't have this, we'll need to either make our own interface, based on [responselike](https://github.com/lukechilds/responselike) or use responselike's own `Response`.

fnky commented on source/merge.ts:
```diff
@@ -0,0 +1,85 @@
+	for (const source of sources) {
+		if (source.hooks) {
+			for (const hook of knownHookEvents) {
+				hooks[hook] = hooks[hook].concat(source.hooks[hook] || []);
```
I caught an issue here, where empty arrays passed by the user would otherwise end up being an Array with an `undefined` item. Kudos to TypeScript.

sindresorhus commented on source/known-hook-events.ts:
```diff
@@ -1,35 +1,67 @@
+	/**
+	 * Called with an `Error` instance. The error is passed to the hook right before it's thrown. This is especially useful when you want to have more detailed errors.
+	 *
+	 * @note Errors thrown while normalizing input options are thrown directly and not part of this hook.
```
I don't think there's any `@note` tag. I usually just use `Note: `.

fnky commented on source/merge.ts:
```diff
@@ -60,16 +61,14 @@ type NextFunction = (error?: Error | string) => void;
 
 type IterateFunction = (options: Options) => void;
 
-type Method = 'GET' | 'PUT' | 'HEAD' | 'DELETE' | 'OPTIONS' | 'TRACE';
```
What do you think to allow lowercase too, e.g.

```ts
type Method = 'GET' | 'PUT' | 'HEAD' ... | 'get' | 'put' | 'head' ...
```

fnky replied:
We could, but this then requires TypeScript users to import it every time the need to update `method`. Do you think that would be too cumbersome for this?

szmarczak commented on source/known-hook-events.ts:
```diff
@@ -1,8 +1,81 @@
+
+export type InitHook = (options: Options) => void;
+// TODO: The `Error` type should confirm to any possible extended error type that can be thrown. See https://github.com/sindresorhus/got#hooksbeforeerror
+export type BeforeErrorHook = (error: Error) => Error | void | Promise<Error> | Promise<void>;
```
I'm not sure why the `void` type is here... Shouldn't it be just `Error | Promise<Error>`?

szmarczak commented on source/known-hook-events.ts:
```diff
@@ -1,8 +1,81 @@
+export type BeforeRequestHook = (options: Options) => void | Promise<void>
+export type BeforeRedirectHook = (options: Options) => void | Promise<void>
+export type BeforeRetryHook = (options: Options, error: HTTPError, retryCount: number) => void | Promise<void>
+export type AfterResponseHook = (response: IncomingMessage, retryWithMergedOptions: (options: Options) => IncomingMessage | CancelableRequest<IncomingMessage>) => IncomingMessage | CancelableRequest<IncomingMessage>
```
I think it should be

```diff
-retryWithMergedOptions: (options: Options) => IncomingMessage | CancelableRequest<IncomingMessage>
+retryWithMergedOptions: (options: Options) => CancelableRequest<IncomingMessage> 
```

fnky replied:
I think this may be a confusion with the line as it is currently:

```ts
retryWithMergedOptions: (options: Options) => CancelableRequest<IncomingMessage>
```

I'll make a type for `RetryFunction` instead to make it more readable.

szmarczak commented on source/known-hook-events.ts:
```diff
@@ -1,8 +1,81 @@
+
+/**
+ * Hooks allow modifications during the request lifecycle. Hook functions may be async and are run serially.
+ */
```
@sindresorhus Do we need to double the documentation?

sindresorhus replied:
https://github.com/sindresorhus/got/pull/714#discussion_r251441415

szmarczak replied:
Agreed on that thing. Sorry I didn't shape my question correctly: I mean, generally, do we need to redocument *everything* in the code, isn't the README just enough?

fnky replied:
I think the issue with using the README is that you quickly end up with out of date documentation.

For example in the documentation for [`beforeError`](https://github.com/sindresorhus/got/#hooksbeforeerror) hook, the example uses `onError` as the property name—this could be a typo or a change that didn't get updated, but nevertheless it makes it more difficult.

If we redocument everything, we can later generate the documentation.

szmarczak replied:
> If we redocument everything, we can later generate the documentation.

Looks promising! Let's do that way then :)

szmarczak commented on source/merge.ts:
```diff
@@ -0,0 +1,85 @@
+
+export function mergeOptions(...sources: Options[]) {
+	sources = sources.map(source => source || {});
+	const merged = merge({}, ...sources);
```
The type of `merged` should be `Options`

fnky replied:
Added explicit return type for `merged` function.

szmarczak commented on source/merge.ts:
```diff
@@ -0,0 +1,93 @@
+		}
+	}
+
+	merged.hooks = hooks as Hooks;
```
[This would fix that](https://github.com/sindresorhus/got/pull/714/files#r251524354)

szmarczak commented on source/utils/types.ts:
```diff
@@ -0,0 +1,21 @@
+	decompress?: boolean;
+	encoding?: BufferEncoding | null;
+	method?: Method;
+	[ key: string ]: unknown | Options;
```
Is the `Options` type needed here?

sindresorhus replied:
We should also add a comment that the line should be removed in the future when the TS migration is complete.

fnky commented:
I have addressed some of the comments :)

sindresorhus commented:
Thank you, @fnky! This is excellent 🙌
